sekoia_automation/connector/event_aggregator.py

```diff
@@ -0,0 +1,187 @@
+import copy
+import datetime
+import threading
+import time
+from typing import Callable, Tuple
+
+import xxhash as xxhash
+from ciso8601 import parse_datetime_as_naive
+from pydantic import BaseModel
+
+
+class Fingerprint(BaseModel):
+    condition_func: Callable[[dict], bool]
+    build_hash_str_func: Callable[[dict], str]
+    ttl: int
+
+
+class Aggregation(BaseModel):
+    start: datetime.datetime
+    end: datetime.datetime
+    ttl: int
+    count: int
+    event: dict
+    fingerprint: str
+
+    def get_aggregated_event(self):
+        aggregated_event = copy.deepcopy(self.event)
+        aggregated_event["sekoiaio"]["repeat"] = {
+            "start": self.start.isoformat(),
+            "end": self.end.isoformat(),
+            "count": self.count,
+            "fingerprint": self.fingerprint,
+        }
+        return aggregated_event
+
+
+class EventAggregatorTTLThread(threading.Thread):
+    f_must_stop: bool  # thread will stop if this flag is active
+    ttl: int  # ttl value
+    on_flush_func: Callable[
+        [dict], None
+    ]  # function to be call when aggregation is flushed
+
+    def __init__(
+        self,
+        event_aggregator: "EventAggregator",
+        on_flush_func: Callable[[dict], None],
+        delay: float = 10,
+    ):
+        super().__init__()
+        self.event_aggregator = event_aggregator
+        self.f_must_stop = False
+        self.on_flush_func = on_flush_func
+        self.delay = delay
+
+    def stop(self) -> None:
+        self.f_must_stop = True
+
+    def run(self) -> None:
+        try:
+            while not self.f_must_stop:
+                time.sleep(self.delay)
+                for aggregated_event in self.event_aggregator.flush_all_ttl():
+                    self.on_flush_func(aggregated_event)
+        finally:
+            for aggregated_event in self.event_aggregator.flush_all():
+                self.on_flush_func(aggregated_event)
+
+
+class EventAggregator:
+    ttl_thread: EventAggregatorTTLThread | None
+    lock: threading.Lock
+
+    def __init__(self, aggregation_definitions: dict[str, list[Fingerprint]]):
+        self.aggregation_definitions = aggregation_definitions
+        self.aggregations: dict[str, Aggregation] = dict()
+        self.lock = threading.Lock()
```

Comment:
This approach looks very interesting.
The main issue that prevented me from implementing on this initial version is the complexity that comes with the following expectation that every shard must be handled by dedicated threads to prevent one locked shard to block the others.

But obviously, if you believe this current version will have a major impact on the performances of a collector I will implement this sharding mecanism in this PR.

Reply:
I agree with you, let's not add complexity that may not even bee needed.

It was just a thought I got while reviewing your code :)